lib/erlef/session.ex

```diff
@@ -0,0 +1,201 @@
+defmodule Erlef.Session do
+  @moduledoc """
+  Erlef.Session - Session handler and session serializer/deserializer
+  """
+
+  @enforce_keys [:account_id, :access_token, :refresh_token, :username, :expires_at, :is_admin]
+  defstruct [:account_id, :access_token, :refresh_token, :username, :expires_at, :is_admin]
+
+  @type auth_data() :: map()
+
+  @type t() :: %__MODULE__{
+          account_id: integer(),
+          access_token: String.t(),
+          refresh_token: String.t(),
+          username: String.t(),
+          expires_at: String.t(),
+          is_admin: boolean()
+        }
+
+  @data_map %{
+    "account_id" => :account_id,
+    "access_token" => :access_token,
+    "refresh_token" => :refresh_token,
+    "username" => :username,
+    "expires_at" => :expires_at,
+    "is_admin" => :is_admin
+  }
+
+  @spec new(String.t(), map()) :: t()
+  def new(username, %{"expires_in" => expires_in} = data) do
+    %{
+      "Permissions" => [%{"AccountId" => aid} | _],
+      "access_token" => access_token,
+      "refresh_token" => refresh_token
+    } = data
+
+    case is_admin(aid, access_token) do
+      truth when is_boolean(truth) ->
+        %__MODULE__{
+          account_id: aid,
+          access_token: access_token,
+          refresh_token: refresh_token,
+          username: username,
+          expires_at: expires_at(expires_in),
+          is_admin: truth
+        }
+
+      err ->
+        err
+    end
+  end
+
+  @spec normalize(map()) :: t() | nil
+  def normalize(nil), do: nil
+
+  def normalize(data) do
+    normal =
+      Enum.reduce(data, %{}, fn {k, v}, acc ->
+        case @data_map[k] do
+          nil ->
+            acc
+
+          key ->
+            Map.put(acc, key, v)
+        end
+      end)
+
+    struct(__MODULE__, normal)
```

Comment:
If this is the re-serialization, you will want to explicitly unset the `is_admin` value and re-check it every time, unless you're able to sign the payload. Otherwise anyone can just set the `is_admin` flag by hand and grant themselves access. 

If `struct/2` does forward to `new/2` you can disregard this (I don't recall the exact rules) but the implementation of `new()` does not make it sound like it'd be doing that automatically since the type of struct to cast has to be passed...

Reply:
The cookie is encrypted and signed, but checking for admin rights is the better way to go and if we have had a db behind us this would be less of an issue, but one problem we face right now is API rate limiting. 

`struct/2` does not forward to new/2, new only gets called by login and refresh, and really could be a private function, I wasn't sure of the shape of things overall when I hammered this out... but devoid of that detail, the flow is this:

- Plug verifies the signature of the encrypted cookie as can be seen here https://github.com/elixir-plug/plug/blob/231004a94f897f80a9199be2b818126ab6885734/lib/plug/session/cookie.ex#L84
- Plug verifies the signature of the payload and decrypts (Erlef.Token is not in play yet)
- incoming request, Plug calls Erlef.Sessions.decode/1 to de-serialize from json to a map (with string keys) which is the "session"
- Further down the chain ErlefWeb.Plug.Session is called. Here we get the session in map form, if the session is not expired, etc. we normalize it to a struct, and store it as "member_session" for the duration of the request. This is is a "nice to have" vs a "need" but it is good in that if there are missing keys, we will blow up
- At the end of the request Plug calls Erlef.Session.encode/1 to serialize back to json and and then encrypts and signs the payload. 

Erlef.Token is not in play yet, I will comment on that below.

Reply:
Yeah with the signature check then it should at least prevent people from just fiddling with the contents and ought to be safe.